validation/markets-validator.js

```diff
@@ -0,0 +1,549 @@
+
+	this.file = file;
+	this.errorsCount = 0;
+	this.warningsCount = 0;
+
+	this.validate = function() {
+		var data = fs.readFileSync(this.file, 'utf8');
+		var json = JSON.parse(data);
+		var features = json.features;
+		var cityName = this.getCityName();
+
+		var featuresValidator = new FeaturesValidator(features, cityName);
+		featuresValidator.validate();
+		featuresValidator.printIssues();
+		this.errorsCount += featuresValidator.getErrorsCount();
+		this.warningsCount += featuresValidator.getWarningsCount();
+
+		var metadataValidator = new MetadataValidator(json.metadata);
+		metadataValidator.validate();
+		metadataValidator.printWarnings();
+		metadataValidator.printErrors();
+		this.errorsCount += metadataValidator.getErrorsCount();
+		this.warningsCount += metadataValidator.getWarningsCount();
+
+		this.printSummary();
+	};
+
+	this.getCityName = function() {
+		return path.basename(this.file, path.extname(this.file));
+	};
+
+
+	this.printSummary = function() {
+		if (this.errorsCount === 0) {
+			this.printSuccess();
+
+		} else {
+			this.printFailure();
+			exitCode = 1;
+		}
+	};
+
+	this.printSuccess = function() {
+		if (this.warningsCount === 0) {
+			console.log("\nValidation PASSED without warnings or errors.".passed);
+		} else {
+			console.log("\nValidation PASSED with %d warning(s) and no errors.".passed, this.warningsCount);
+		}
+	};
+
+	this.printFailure = function() {
+		console.log("\nValidation done. %d warning(s), %d error(s) detected.".error, this.warningsCount, this.errorsCount);
+	};
+
+}
+
+
+function FeaturesValidator(features, cityName) {
+
+	this.features = features;
+	this.cityName = cityName;
+	this.errorsCount = 0;
+	this.warningsCount = 0;
+	this.featureValidators = [];
+
+	this.validate = function() {
+		for (var i = 0, length = features.length; i < length; ++i) {
+			var feature = features[i];
+			this.validateFeature(feature);
+		}
+	};
+
+	this.validateFeature = function(feature) {
+		var featureValidator = new FeatureValidator(feature, this.cityName);
+		featureValidator.validate();
+		if (featureValidator.hasErrors()) {
+			this.errorsCount += featureValidator.getErrorsCount();
+		}
+		if (featureValidator.hasWarnings()) {
+			this.warningsCount += featureValidator.getWarningsCount();
+		}
+		this.featureValidators.push(featureValidator);
+	};
+
+	this.getErrorsCount = function() {
+		return this.errorsCount;
+	};
+
+	this.getWarningsCount = function() {
+		return this.warningsCount;
+	};
+
+	this.printIssues = function() {
+		for (var i = 0, length = this.featureValidators.length; i < length; ++i) {
+			var featureValidator = this.featureValidators[i];
+			featureValidator.printWarnings();
+			featureValidator.printErrors();
+		}
+	};
+
+}
+
+
+function FeatureValidator(feature, cityName) {
+
+	this.feature = feature;
+	this.cityName = cityName;
+	this.errors = [];
+	this.warnings = [];
+
+	this.validate = function() {
+		var feature = this.feature;
+		if (feature === undefined) {
```

Comment:
Maybe extract all the `===undefined`, `===null`, `==={}` calls into a helper function? Seeing that this is done in `validate`, `validateProperties`, and `validateTitle`